### PR TITLE
MAILBOX-304 BlobId migration for attachments

### PR DESCRIPTION
**Pull request by chibenwa**
Only the 3 last commits

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java [hunk 1]
```diff
@@ -48,16 +52,21 @@
 
 public class CassandraAttachmentDAO {
 
+    public static final int SELECT_ALL_TIMEOUT = toIntExact(TimeUnit.DAYS.toMillis(1));
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java]: what about making it configurable? If it's not sufficient, I would like not to have to wait also for a merge request to change it

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java [hunk 2]
```diff
@@ -65,10 +67,15 @@ public CassandraAttachmentDAO(Session session, CassandraUtils cassandraUtils) {
 
         this.selectStatement = prepareSelect(session);
         this.selectAllStatement = prepareSelectAll(session);
+        this.deleteStatement = prepareDelete(session);
         this.insertStatement = prepareInsert(session);
         this.cassandraUtils = cassandraUtils;
     }
 
+    private PreparedStatement prepareDelete(Session session) {
+        return session.prepare(delete().from(TABLE_NAME).where(eq(ID, bindMarker(ID))));
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java]: indent

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2MigrationTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        migration.run();
+
+        assertThat(attachmentDAOV2.getAttachment(ATTACHMENT_ID).join())
+            .contains(CassandraAttachmentDAOV2.from(attachment1, BlobId.forPayload(attachment1.getBytes())));
+        assertThat(attachmentDAOV2.getAttachment(ATTACHMENT_ID_2).join())
+            .contains(CassandraAttachmentDAOV2.from(attachment2, BlobId.forPayload(attachment2.getBytes())));
+    }
+
+    @Test
+    public void migrationShouldRemoveAttachmentsFromV() throws Exception {
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2MigrationTest.java]: *V1

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2MigrationTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        attachmentDAO.storeAttachment(attachment2).join();
+
+        migration.run();
+
+        assertThat(attachmentDAO.getAttachment(ATTACHMENT_ID).join())
+            .isEmpty();
+        assertThat(attachmentDAO.getAttachment(ATTACHMENT_ID_2).join())
+            .isEmpty();
+    }
+}
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2MigrationTest.java]: not test about incomplete migration?

**Comment by chibenwa** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java [hunk 5]
```diff
@@ -83,6 +97,15 @@ private PreparedStatement prepareSelect(Session session) {
             .thenApply(optional -> optional.map(this::attachment));
     }
 
+    public Stream<Attachment> retrieveAll() {
+        return cassandraUtils.convertToStream(
+            cassandraAsyncExecutor.execute(
+            selectAllStatement.bind()
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java]: indent

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java [hunk 6]
```diff
@@ -42,9 +45,14 @@ public AttachmentV2Migration(CassandraAttachmentDAO attachmentDAOV1,
...
     public MigrationResult run() {
-        return attachmentDAOV1.retrieveAll()
-            .map(this::migrateAttachment)
-            .reduce(MigrationResult.COMPLETED, Migration::combine);
+        try {
+            return attachmentDAOV1.retrieveAll()
+                .map(this::migrateAttachment)
+                .reduce(MigrationResult.COMPLETED, Migration::combine);
+        } catch (Exception e) {
+            LOGGER.error("Error while performing attachment V2 migration", e);
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java]: V2 is not very clear here: is it schema V2, attachmentDAO V2?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java [hunk 7]
```diff
@@ -56,6 +64,7 @@ private MigrationResult migrateAttachment(Attachment attachment) {
                 .join();
             return MigrationResult.COMPLETED;
         } catch (Exception e) {
+            LOGGER.error("Error while performing attachment V2 migration", e);
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/AttachmentV2Migration.java]: idem

**Comment by chibenwa** [comment]
test this please

**Comment by aduprat** [comment]
Merge: https://github.com/linagora/james-project/pull/975

**Pull request closed without merging**